### DISCUSSION TITLE
데모데이 준비용 터놓고 데모사이트 구성 논의

### DISCUSSION THREAD

**Discussion by soominsohn** [woowacourse-teams/2022-ternoko#468] [General]
**전체적인 흐름 예상**

1. loginCrew, loginCoach 컨트롤러, 서비스 메서드를 생성한다.
2. 프론트에서 크루로그인, 코치로그인으로 접근가능!
3. 그럼 백엔드에서 응답으로 크루, 코치에 대한 jwt token 발급!
4. token만 있으면 서비스 이용 가능

즉, 로그인을 mock해준다고 생각하면 좋다!

**[git 사용 브랜치]**

**demo 브랜치에서 진행하면 좋을 것 같아용!**

[**DB 전처리]**

1. member table
    - 단, 1~10은 인터뷰 요청 가능한 coach로 넣어 놓는다.
    - id 10~109 코치 더미데이터
    - id 110 ~ 209 크루 더미데이터

사용자 요청이 들어오면

코치: 10~109별로 순차적으로 로그인 가능

크루: 110~209별로 순차적으로 로그인 가능

1. available_date_time table
    1. 1~10번 코치에 대해 available_date_time 미리 열어두기 

**[백엔드 코드단 요구사항]**

1. AuthController, Service 안에 loginCrew, loginCoach 메서드를 생성한다.
    1. loginCoach 서비스 메서드 
        1. private static int COACH_ID를 선언해 놓는다!
        2. 요청 들어올 때마다 ++ 시키고, DB에서 찾을 때 id값 (COACH_ID) %100 + 10 하면 될듯!
    2. loginCrew 서비스 메서드 
        1. private static int CREW_ID를 선언해 놓는다!
        2. 요청 들어올 때마다 ++ 시키고, DB에서 찾을 때 id값 (CREW_ID) %100 + 110 하면 될듯!
    
2. 해당 컨트롤러로 요청보내면 access token 발급받아진다!
3. 이제 마음껏 요청 날려보면됨!
4. email, slack 사용부분은 코드내에서 주석처리 해버린다!
    1. 워낙 코드단에 의존성 분리 잘 해놓으셔서 뚝딱 가능하네용~ㅎㅎ!!

**[프론트엔드 요구사항]**

<img width="1340" alt="image" src="https://user-images.githubusercontent.com/83059234/195510640-e7f3dbe0-6571-44d6-898d-0b1de2a024d6.png">

1. 슬랙 로그인 부분에
    1. 코치 로그인 버튼 (GET “/api/login/coach”)
    2. 크루 로그인 버튼 (GET “/api/login/crew”)
    
    추가해주실 수 있나용~? 😄
    
    해당 로그인 버튼 누르면 로그인처럼
    
    access token 발급받고 코치 메인페이지/ 크루 메인페이지로 이동할 수 있게끔…
    

**[인프라 요구사항]**

요약: EC2 두개 파고, PORD와 DEV 인프라 구성이랑 다른 구성으로 갈거에용! (최소한으로 가고싶어서..ㅎㅎ)

1. nginx + webserver http://demo.ternoko.site 프론트 도메인 생성)
    1. 프론트 정적파일 넣어놓기
    2. http://demo.ternoko.site/api/~~~ 요청 백엔드 WAS로 넘겨주기
2. WAS (Spring + DB 같은 인스턴스에 생성할게요)

@lokba @sanaandmomo @her0807 @Juhyung990122 @HyeonbinSa @dongho108 @soominsohn 
혹시 이런 구성 괜찮으신지, 다른 의견 있으신지 피드백 주시면 감사합니다!

**Comment by her0807** [comment]
앤지 빠르게 계획 수립해주셨네요 ! ☺️👍🏻

## 제안 사항 

- 프론트에서 크루 버튼을 누르면 api에 `demo/login/crew`, 요청을 보낸다. 
   -  백엔드에서 new Crew 생성해서 토큰 발급해서 반환한다. 
 
- 코치 버튼을 누르면 api에 `demo/login/coach` 요청을 보낸다. 
   -   백엔드에서 new Coach 생성해서 토큰 발급해서 반환한다. 

---



[git 사용 브랜치]

demo 브랜치에서 진행하면 좋을 것 같아용! -> 👍🏻


available_date_time table
1~10번 코치에 대해 available_date_time 미리 열어두기 👍🏻
[백엔드 코드단 요구사항]


이제 마음껏 요청 날려보면됨!


## 고민 사항
지금 면담 생성 플로우만 타고 , 히스토리나 코멘트는 못봐서 아쉬운데.. 
인터뷰 완료 상태로 만들기 ! 버튼 누르면 완료된 면담으로 보내버리는 부분 하나 추가할까요 ?

**Reply by dongho108** [reply]
강제로 상태바꾸는거 추가하면 사이드 이펙트 없을까요..? 시간이 이상하게 뜬다던지.. 원래는 시간이 지나야 면담완료인데요!

**Reply by soominsohn** [reply]
오홍 히스토리나 코멘트 볼 수 있게 얘도 더미데이터로 넣어놓으면 어때용~?

**Reply by HyeonbinSa** [reply]
수달이 제안하신 내용 중에 api를 바꾸는거에 찬성입니다!!

코치나 크루나 현재 더미데이터를 넣는 것이니 히스토리나 코멘트 또한 더미데이터로 넣어놓는 것은 어떨까생각합니다.
`인터뷰 완료 상태로 만들기` 같은 경우 추가적인 구현이 더 필요해보이는데 더미데이터로 대체하는게 좋을 것 같습니다. (저희가 실 서비스에는 사용하지 않는 기능이도 하구요. 데모 때는 있었지만 실 사용에는 완료하기 버튼이 없다면 의아해할 수도 있다는 생각이 들었습니다.)

**Comment by dongho108** [comment]
설계하느라 고생하셨어요 앤지~

몇가지 의문점이 들어서 남겨볼게요

1. controller(싱글톤)에 static 변수를 둬도 될까요?
2. 코치수나 회원수를 정해서 DB에 저장해 놓을 필요가 있을까요?
   - 로그인 요청 올때마다 새로운 user로 만들어주는게 더 나을 것 같아요~ 
   - 그러면 controller에 static으로 굳이 유저수를 세지 않아도 될 것 같아요


## 제가 생각하는 예상 시나리오는

10명의 더미 코치가 있다.

1. 프론트 페이지에서 코치 로그인 or 크루 로그인을 누르면
2. 각 role에 맞게 새로운 유저를 생성해서 넣어줍니다.
3. 각 role에 맞게 테스트해볼 수 있다.
   - 크루 로그인시 : 10명의 더미코치만 볼 수 있다, 면담 신청 가능
   - 코치 로그인시 : 시간을 열어보는 경험만 할 수 있다, 면담 내역은 확인 못함(데이터가 없어서)
 
**제 방식의 문제점**
- 코치 사용 경험은 면담을 여는 경험밖에 못한다. 스케줄(달력)에 면담 내용들을 볼 수 없다..ㅠ



## 추가로 걱정되는점

설계 보니까 시간이 많이 걸릴 수 있는 작업인 것 같아서.. 이슈 분리하고 분업하는게 어떨까 싶네요!!



## 피드백
제 방식에 대해 궁금하거나, 예상되는 문제점, 혹은 제가 적어놓은 문제점 해결방법이 있다면 답글 주시면 감사하겠습니다~~

**Reply by soominsohn** [reply]
2. 각 role에 맞게 새로운 유저를 생성해서 넣어줍니다.  `-> (가능할 것 같기도...?)`

코치 사용 경험은 면담을 여는 경험밖에 못한다. 스케줄(달력)에 면담 내용들을 볼 수 없다..ㅠ
`-> 지난 내역들을 상태별로 더미데이터로 넣어놓으면 어떨까요~?
`

**Reply by dongho108** [reply]
새로운 유저한테 면담내역을 미리 넣을 수 있을까요??

**Reply by HyeonbinSa** [reply]
애쉬가 말씀해주신대로 버튼을 누를 때마다 코치나 크루를 생성하는 것도 방법이겠네요. static 변수와 상관없이 Auto Increment 되는 id값을 사용하면되니 편할 수도 있을 것 같습니다.

여기서 수달의 의견에 대한 제 생각을 적다 생각난건데 사용자들에게 저희 서비스에 대해 보여주기 위해 완료된 면담이나 열려있는 시간, 코멘트 등을 보여준다고 가정했을 때 코치나 크루 데이터를 미리 만들어두는게 좋을 것 같다고 생각했습니다.

실제 운영 환경이라면 싱글톤에서 static을 사용했을 때 문제가 있겠지만 데모 버전 환경에서는 괜찮다고 생각합니다.
그리고 Interview 내용처럼 공유되는 내용도 아닐 뿐더러 사용되는 곳이 login() 이라는 메서드에 한해 사용되니 문제없어보입니다.

**Comment by her0807** [comment]
추가로 다음주까지 시간이 있을 것 같으니까 우려되는 점들 여러개 적어두고 해결방안 같이 찾아보면 좋을 듯 싶습니다

**Reply by soominsohn** [reply]
맞아용! 계속 추가해주시면 같이 고민해봐용

**Comment by HyeonbinSa** [comment]
앤지 데모 사이트를 구현하시느라 고생 많으셨습니다.
테스트해보기도 하고 오전에 회의했던 내용대로 구현이 된 것 같아서 저는 현재 구현 상태가 좋다고 생각합니다.
